From c9fde86b0a2440133bc08f4811b6ca793be47f0a Mon Sep 17 00:00:00 2001
From: Volker Hilsheimer <volker.hilsheimer@qt.io>
Date: Thu, 11 Mar 2021 14:01:28 +0100
Subject: [PATCH 254/254] Revert "QPushButton: fix support of style sheet rule
 for text alignment"

This reverts commit 6269438af95bbd988ead08829fa3bc9dc25891e8, and adds a test.

This change introduced QTBUG-91735, without fixing QTBUG-86857 correctly. The
code already interprets the textAlignment values from the rule, also if no
icon is set. Adding the same, or some default textAlignment to the text flags
if there is no icon doesn't work.

Fixes: QTBUG-91735
Task-number: QTBUG-86857
Pick-to: 6.1 6.0 5.15
Change-Id: Iee07e63a40e72909275f32e1caa28b33a595f879
Reviewed-by: Oliver Eftevaag <oliver.eftevaag@qt.io>
Reviewed-by: Richard Moe Gustavsen <richard.gustavsen@qt.io>
(cherry picked from commit 674747bac1c8b57d4940de2ee68b6b562dfcad61)
---
 src/widgets/styles/qstylesheetstyle.cpp        |  6 ++----
 .../widgets/qpushbutton/tst_qpushbutton.cpp    | 18 ++++++++++++++++++
 2 files changed, 20 insertions(+), 4 deletions(-)

diff --git a/src/widgets/styles/qstylesheetstyle.cpp b/src/widgets/styles/qstylesheetstyle.cpp
index 72f7ad7455..822e6f895f 100644
--- a/src/widgets/styles/qstylesheetstyle.cpp
+++ b/src/widgets/styles/qstylesheetstyle.cpp
@@ -3540,8 +3540,8 @@ void QStyleSheetStyle::drawControl(ControlElement ce, const QStyleOption *opt, Q
                 const uint horizontalAlignMask = Qt::AlignHCenter | Qt::AlignLeft | Qt::AlignRight;
                 const uint verticalAlignMask = Qt::AlignVCenter | Qt::AlignTop | Qt::AlignLeft;
 
-                const Qt::Alignment textAlignment = rule.position()->textAlignment;
-                if (rule.hasPosition() && textAlignment != 0) {
+                if (rule.hasPosition() && rule.position()->textAlignment != 0) {
+                    Qt::Alignment textAlignment = rule.position()->textAlignment;
                     tf |= (textAlignment & verticalAlignMask) ? (textAlignment & verticalAlignMask) : Qt::AlignVCenter;
                     tf |= (textAlignment & horizontalAlignMask) ? (textAlignment & horizontalAlignMask) : Qt::AlignHCenter;
                     if (!styleHint(SH_UnderlineShortcut, button, w))
@@ -3600,8 +3600,6 @@ void QStyleSheetStyle::drawControl(ControlElement ce, const QStyleOption *opt, Q
                         iconRect.translate(pixelMetric(PM_ButtonShiftHorizontal, opt, w),
                                            pixelMetric(PM_ButtonShiftVertical, opt, w));
                     p->drawPixmap(iconRect, pixmap);
-                }else {
-                    tf |= textAlignment;
                 }
 
                 if (button->state & (State_On | State_Sunken))
diff --git a/tests/auto/widgets/widgets/qpushbutton/tst_qpushbutton.cpp b/tests/auto/widgets/widgets/qpushbutton/tst_qpushbutton.cpp
index 94cb42cc00..e818514a79 100644
--- a/tests/auto/widgets/widgets/qpushbutton/tst_qpushbutton.cpp
+++ b/tests/auto/widgets/widgets/qpushbutton/tst_qpushbutton.cpp
@@ -68,6 +68,7 @@ private slots:
     void taskQTBUG_20191_shortcutWithKeypadModifer();
     void emitReleasedAfterChange();
     void hitButton();
+    void iconOnlyStyleSheet();
 
 protected slots:
     void resetCounters();
@@ -695,5 +696,22 @@ void tst_QPushButton::hitButton()
     QVERIFY(!button2->hitButton(QPoint(2, 2)));
 }
 
+/*
+    Test that a style sheet with only icon doesn't crash.
+    QTBUG-91735
+*/
+void tst_QPushButton::iconOnlyStyleSheet()
+{
+    QIcon icon(":/qt-project.org/styles/commonstyle/images/dvd-32.png");
+    QVERIFY(!icon.isNull());
+    QPushButton pb;
+    pb.setStyleSheet("QPushButton {"
+        "icon: url(:/qt-project.org/styles/commonstyle/images/dvd-32.png);"
+        "border: red;"
+    "}");
+    pb.show();
+    QVERIFY(QTest::qWaitForWindowExposed(&pb));
+}
+
 QTEST_MAIN(tst_QPushButton)
 #include "tst_qpushbutton.moc"
-- 
2.33.1

